From e66d23f3a127dbad70b4534c5405b08b7fc4c3c7 Mon Sep 17 00:00:00 2001
From: Bin Han <bin.b.han@intel.com>
Date: Wed, 3 Dec 2014 13:07:42 +0800
Subject: [PATCH 176/176] atomisp2: keep the parameter not updated until
 atomisp_load_acc_extension

Because atomisp_load_acc_extension will destroy stream/pipes and re-create
them in order the extended binary can be loaded to pipe, this will cause ISP
parameter lost which was configured before that. The patch just changes that
before really starting the stream, ISP parameters are stored and a flag is
used to keep the update need, and after atomisp_load_acc_extension and before
stream starts, check the update flag and update ISP parameters.

Change-Id: I02ffa8549fda73a5912f393678e0afdb52ff7c16
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-3427
Signed-off-by: Bin Han <bin.b.han@intel.com>
Signed-off-by: Tommi Franttila <tommi.franttila@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/31262
Reviewed-by: Jukka O Kaartinen <jukka.o.kaartinen@intel.com>
Approver: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
---
 .../camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c  | 3 ++-
 .../drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c       | 6 ++++++
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
index c9504bf..575a932 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
@@ -3470,7 +3470,8 @@ int atomisp_set_parameters(struct video_device *vdev,
 		/* indicate to CSS that we have parameters to be updated */
 		asd->params.css_update_params_needed = true;
 
-		if (asd->stream_env[ATOMISP_INPUT_STREAM_GENERAL].stream) {
+		if (asd->stream_env[ATOMISP_INPUT_STREAM_GENERAL].stream &&
+		    asd->streaming == ATOMISP_DEVICE_STREAMING_ENABLED) {
 			atomisp_css_update_isp_params(asd);
 			asd->params.css_update_params_needed = false;
 		}
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
index 9840049..925d9de 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_ioctl.c
@@ -1765,6 +1765,12 @@ static int atomisp_streamon(struct file *file, void *fh,
 		dev_err(isp->dev, "acc extension failed to load\n");
 		goto out;
 	}
+
+	if (asd->params.css_update_params_needed) {
+		atomisp_css_update_isp_params(asd);
+		asd->params.css_update_params_needed = false;
+	}
+
 	ret = atomisp_css_start(asd, css_pipe_id, false);
 	if (ret)
 		goto out;
-- 
1.9.1

